docs/plugins/discovery-ec2.asciidoc

```diff
@@ -1,167 +1,215 @@
+{ref}/modules-discovery-hosts-providers.html[discovery process] by querying the
+https://github.com/aws/aws-sdk-java[AWS API] for a list of EC2 instances
+matching certain criteria determined by the <<discovery-ec2-settings,plugin
+settings>>.
 
-*If you are looking for a hosted solution of Elasticsearch on AWS, please visit http://www.elastic.co/cloud.*
+*If you are looking for a hosted solution of Elasticsearch on AWS, please visit
+http://www.elastic.co/cloud.*
 
 :plugin_name: discovery-ec2
 include::install_remove.asciidoc[]
 
 [[discovery-ec2-usage]]
-==== Getting started with AWS
+[[discovery-ec2-settings]]
+==== Using the EC2 discovery plugin
 
-The plugin adds a seed hosts provider named `ec2`. This seed hosts provider
-finds other Elasticsearch instances in EC2 by querying the AWS metadata
-service.  Authentication is done using
-http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html[IAM
-Role] credentials by default. To enable the plugin, configure {es} to use the
-`ec2` seed hosts provider:
+The `discovery-ec2` plugin adds a seed hosts provider named `ec2` which
+provides a list of seed addresses to the
+{ref}/modules-discovery-hosts-providers.html[discovery process] by querying the
+https://github.com/aws/aws-sdk-java[AWS API] for a list of EC2 instances
+matching certain criteria including their tags, membership of certain security
+groups, and placement in certain availability zones. The interaction with the
+AWS API can be authenticated using an
+http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html[instance
+or container role], or else custom credentials can be supplied.
+
+===== Enabling EC2 discovery
+
+To enable EC2 discovery, configure {es} to use the `ec2` seed hosts provider:
 
 [source,yaml]
 ----
 discovery.seed_providers: ec2
 ----
 
-==== Settings
+===== Configuring EC2 discovery
 
-EC2 discovery supports a number of settings.  Some settings are sensitive and
-must be stored in the {ref}/secure-settings.html[elasticsearch keystore].  For
-example, to use explicit AWS access keys:
+EC2 discovery supports a number of settings. Some settings are sensitive and
+must be stored in the {ref}/secure-settings.html[{es} keystore]. For example,
+to authenticate using a particular access key and secret key, add these keys to
+the keystore by running the following commands:
 
 [source,sh]
 ----
 bin/elasticsearch-keystore add discovery.ec2.access_key
 bin/elasticsearch-keystore add discovery.ec2.secret_key
 ----
 
-The following are the available discovery settings. All should be prefixed with `discovery.ec2.`.
-Those that must be stored in the keystore are marked as `Secure`.
+The available settings for the EC2 discovery plugin are as follows.
+
+`discovery.ec2.access_key`::
+
+    An EC2 access key. If set, you must also set `discovery.ec2.secret_key`.
+    If unset, `discovery-ec2` will instead use the instance or container role
```

Comment:
do we have tests for this?

Reply:
No, I don't think so. We use the `DefaultAWSCredentialsProviderChain` which has this behaviour, and as noted in https://github.com/elastic/elasticsearch/pull/19561#issuecomment-235927422 we should have tests for this (see also internal issues linked near to that about such tests)

Reply:
This is unfortunately completely untested, and based on a preliminary investigation, does not work. The repository-s3 plugin has some special handling to make some of this work and has extra testing, but the ec2-discovery plugin lacks this, and also the SDK version used there does not give us the capabilities to test...
Let's revert this back to the old text for now.

Reply:
On reflection I'm not sure this plugin is the right way to do discovery on ECS/EKS anyway, but I'd be interested to hear why this doesn't work.

I removed mention of container roles in 19de646c56c6e3199ec593887affc2e44e1ca668 but did not fully revert to the old text as this PR fixes other issues with it.

Reply:
I've opened https://github.com/elastic/elasticsearch/pull/41732 which will hopefully provide some clarification on this.

Reply:
We can now update this with the findings in #41732